RecoEgamma/EgammaTools/src/EnergyScaleCorrection.cc

```diff
@@ -0,0 +1,431 @@
+
+#define LOGVERB(x) LogTrace(x)
+#define LOGWARN(x) edm::LogWarning(x)
+#define LOGERR(x) edm::LogError(x)
+#define LOGDRESSED(x) LogDebug(x)
+
+EnergyScaleCorrection::EnergyScaleCorrection(const std::string& correctionFileName, unsigned int genSeed):
+  smearingType_(ECALELF)
+{
+
+  if(!correctionFileName.empty()) { 
+    std::string filename = correctionFileName+"_scales.dat";
+    readScalesFromFile(filename);
+    if(scales_.empty()) {
+      throw cms::Exception("EnergyScaleCorrection") << "[ERROR] scale correction map empty";
+    }
+  }
+  
+  if(!correctionFileName.empty()) { 
+    std::string filename = correctionFileName+"_smearings.dat";
+    readSmearingsFromFile(filename);
+    if(smearings_.empty()) {
+      throw cms::Exception("EnergyScaleCorrection") << "[ERROR] smearing correction map empty";
+    }
+  }
+
+}
+
+float EnergyScaleCorrection::scaleCorr(unsigned int runNumber, double et, double eta, double r9,
+				       unsigned int gainSeed, std::bitset<kErrNrBits> uncBitMask) const
+{
+  const ScaleCorrection* scaleCorr =  getScaleCorr(runNumber, et, eta, r9, gainSeed);
+  if(scaleCorr!=nullptr) return scaleCorr->scale();
+  else return kDefaultScaleVal_;
+}
+
+
+
+float EnergyScaleCorrection::scaleCorrUncert(unsigned int runNumber, double et, double eta, double r9,
+					     unsigned int gainSeed, std::bitset<kErrNrBits> uncBitMask) const
+{
+  
+  const ScaleCorrection* scaleCorr = getScaleCorr(runNumber, et, eta, r9, gainSeed);
+  if(scaleCorr!=nullptr) return scaleCorr->scaleErr(uncBitMask);
+  else return 0.;
+}
+
+
+float EnergyScaleCorrection::smearingSigma(int runnr, double et, double eta, double r9,
+					   unsigned int gainSeed, ParamSmear par, 
+					   float nSigma) const
+{
+  if (par == kRho) return smearingSigma(runnr, et, eta, r9, gainSeed, nSigma, 0.);
+  if (par == kPhi) return smearingSigma(runnr, et, eta, r9, gainSeed, 0., nSigma);
+  return smearingSigma(runnr, et, eta, r9, gainSeed, 0., 0.);
+}
+
+
+float EnergyScaleCorrection::smearingSigma(int runnr, double et, double eta, double r9,
+					   unsigned int gainSeed, float nrSigmaRho, 
+					   float nrSigmaPhi) const
+{
+  const SmearCorrection* smearCorr = getSmearCorr(runnr,et,eta,r9,gainSeed);
+						  
+  if(smearCorr!=nullptr) return smearCorr->sigma(nrSigmaRho,nrSigmaPhi);
+  else return kDefaultSmearVal_;
+}
+
+
+const EnergyScaleCorrection::ScaleCorrection* 
+EnergyScaleCorrection::getScaleCorr(unsigned int runnr, double et, double eta, double r9,
+				    unsigned int gainSeed) const
+{
+
+  // buld the category based on the values of the object
+  CorrectionCategory category(runnr, et, eta, r9, gainSeed);
+  auto result = std::equal_range(scales_.begin(),scales_.end(),category,Sorter<CorrectionCategory,ScaleCorrection>()); 
+  auto nrFound = std::distance(result.first,result.second);
+  if(nrFound==0){
+    LOGWARN("EnergyScaleCorrection") << "[ERROR] Scale category not found: " << category << " Returning uncorrected value.";
```

Comment:
very rarely you can get eta>3 electrons. So we print a warning and move on. This is more to catch it if it happens repeatedly (which is a config error). This is not ideal but is what we have.

Reply:
Then this is a case for a LogInfo or even LogDebug, because it's normal operating behavior (even more so in phase-2).
Please cleanup.

Reply:
done. Phase-2 wouldnt be a problem here though because phase-2 hopefully would have scale & smearing corrrections defined correctly for its eta regions